File: 7Feb2024.py
class Solution:
    def frequencySort(self, s: str) -> str:
        def idx(c):
            if c >= 'a':
                return ord(c) - ord('a') + 36
            if c >= 'A':
                return ord(c) - ord('A') + 10
            return ord(c) - ord('0')

        def toChar(i):
            # print(i)
            if i >= 36:
                return chr(i - 36 + ord('a'))
            if i >= 10:
                return chr(i - 10 + ord('A'))
            return chr(i + ord('0'))

        fre = [[0, i] for i in range(62)]
        for char in s:
            fre[idx(char)][0] += 1
        fre.sort(key=lambda x: (-x[0], x[1]))
        result = ""
        for freq, index in fre:
            # print(freq,index,toChar(5))
            result += toChar(index)*freq
        return result

File: 7Feb2024.java
import java.util.Arrays;

class Solution {
    public String frequencySort(String s) {
        int[][] fre = new int[62][2];
        for (int i = 0; i < 62; i++)
            fre[i][1] = i;
        for (int i = 0; i < s.length(); i++)
            fre[idx(s.charAt(i))][0]++;
        Arrays.sort(fre, (a, b) -> (a[0] == b[0]) ? b[1] - a[1] : b[0] - a[0]);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 62; i++) {
            // System.out.println(fre[i][0]);
            while (fre[i][0]-- > 0) {
                sb.append(chr(fre[i][1]));
            }
        }
        return sb.toString();
    }

    public int idx(char c) {
        // System.out.println(c);
        if (c >= 'a')
            return c - 'a' + 36;
        if (c >= 'A')
            return c - 'A' + 10;
        return c - '0';
    }

    public char chr(int i) {
        if (i >= 36)
            return (char) (i - 36 + 'a');
        if (i >= 10)
            return (char) (i - 10 + 'A');
        return (char) (i + '0');
    }
}